src/pyk/utils.py

```
@@ -452,8 +452,10 @@ def run_process(
     delta_time = time.time() - start_time
     logger.info(f'Completed in {delta_time:.3f}s with status {res.returncode}: {command}')
 
-    if check:
-        res.check_returncode()
+    if check and res.returncode != 0:
+        sys.stderr.write(f'[ERROR] Running process failed with returncode {res.returncode}:\n    {shlex.join(args)}\n')
+        sys.stderr.flush()
+        sys.exit(res.returncode)
```

@tothtamas28 should I be writing to `stderr` directly or using `logger.error(...)`? I think the later may be more standard I guess?

---

Hmmmm, using the logger causes timestamps to be printed. Going to stick with `stderr` for now.

---

I think the `pyk kompile` tool should just print to `stderr` / `stdout` as these messages are part of normal tool usage.

But `run_process(..., check=True)` should not result in these side effects, as `CalledProcessError` already has all the data:

```python3
proc_res = subprocess.run(['kompile', '--foo'], stderr=subprocess.PIPE)
try:
    proc_res.check_returncode()
except CalledProcessError as err:
    print(err.cmd)
    print(err.returncode)
    print(err.stderr)
```

So the `kompile` function can just let the exception propagate or do some error translation. The CLI can than catch the exception and perform the side effects.

---

Wouldn't we want a standard behaviour of `run_process(...)` across all our invocations of subtools though? I figured that would be a nice feature to have.

---

Especially since the callsite passed `check`, I think we should just go for standard behavior. If the callsite wants to do something specific with teh result, they can pass `check=False` and handle the returned `CompletedProcess` themselves directly, which is what we do in several places.

---

I agree, this uniformity is desirable for the CLI. But this function has other uses as well (downstreams even ones we're not aware of), so in my opinion it'd be better to perform the side effects on the call site.

We can have a function that handles the error:

```python3
def exit_with_err(err: CalledProcessError) -> None:
    sys.stderr.write(f'[ERROR] Running process failed with returncode {err.returncode}:\n    {shlex.join(err.cmd)}\n')
    sys.stderr.flush()
    sys.exit(err.returncode)
```

Then:

```python3
try:
    kompile(...)
except CalledProcessError as err:
    exit_with_err(err)
```